### PR TITLE
The crashy WebKit XPC service causes Safari to Crash on iOS

### PR DESCRIPTION
**Pull request by msaboff**
#### 2fb42dcd321a3f84800f600179f3d9abdd7131f8
<pre>
The crashy WebKit XPC service causes Safari to Crash on iOS
<a href="https://bugs.webkit.org/show_bug.cgi?id=247395">https://bugs.webkit.org/show_bug.cgi?id=247395</a>
rdar://101876374

Reviewed by Chris Dumez.

When calling sendToAllProcessesForSession(), we should ignore crashy WebProcesses.
Also fixed WebContent.Crashy configuration for iOS so that we install the service and don&apos;t
change the code signing flags.

* Source/WebKit/Configurations/WebContentService.Crashy.xcconfig:
* Source/WebKit/UIProcess/WebProcessPool.h:
(WebKit::WebProcessPool::sendToAllProcessesForSession):
* Source/WebKit/UIProcess/WebProcessProxy.h:
(WebKit::WebProcessProxy::isCrashyProcess const):
(WebKit::WebProcessProxy::isCrashyProcess): Deleted.

Canonical link: <a href="https://commits.webkit.org/256434@main">https://commits.webkit.org/256434@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/4863429e1cc7b87aa67f4d7bfc126021e1e004f6

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/95739 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/4993 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/28781 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/105315 "Built successfully") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/5071 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/33751 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/88120 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/101153 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/101400 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/3727 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/82353 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/30782 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/85590 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/87502 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/73611 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/39484 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/19041 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/37177 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/20358 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4446 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/63/builds/41244 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/42995 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/65/builds/43171 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/39609 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | |

### CODE REVIEW

**Review by achristensen07 — CHANGES_REQUESTED** [review]
This seems to be a consequence of not passing isPrewarmed to initializeNewWebProcess(crashyProcessProxy, nullptr); which makes it so !process->isPrewarmed() is always true.  I think that should be done instead, and if that doesn't fix this or can't be done, then we should make a WebProcessProxy::optionalSessionID() that returns a std::optional<PAL::SessionID> and returns nullopt if there's no WebsiteDataStore, then call that from sendToAllRemoteWorkerProcesses instead.

**Comment by msaboff** [comment]
> This seems to be a consequence of not passing isPrewarmed to initializeNewWebProcess(crashyProcessProxy, nullptr); which makes it so !process->isPrewarmed() is always true. I think that should be done instead, and if that doesn't fix this or can't be done, then we should make a WebProcessProxy::optionalSessionID() that returns a std::optionalPAL::SessionID and returns nullopt if there's no WebsiteDataStore, then call that from sendToAllRemoteWorkerProcesses instead.

I'll investigate those changes.

**Comment by szewai** [comment]
If marking crashy process prewarmed does not work, an alternative might be creating an ephemeral WebsiteDataStore for the crashy process (or making a global ephemeral WebsiteDataStore that is used for all crashy processes). Our current implementation assumes a web process either has a valid WebsiteDataStore(session) or is prewarmed (I think there are many checks for this).

**Review by cdumez — CHANGES_REQUESTED** [review]
The existing code is:
```
template<typename T>
void WebProcessPool::sendToAllProcessesForSession(const T& message, PAL::SessionID sessionID)
{
    for (auto& process : m_processes) {
        if (process->canSendMessage() && !process->isPrewarmed() && process->sessionID() == sessionID)
            process->send(T(message), 0);
    }
}
```

Note how it explicitly disregards prewarmed processes. I think we should explicitly ignore crashy processes too instead of returning an invalid ID (which is a new thing I'd rather not have to deal with).

**Comment by msaboff** [comment]
> The existing code is:
> 
> ```
> template<typename T>
> void WebProcessPool::sendToAllProcessesForSession(const T& message, PAL::SessionID sessionID)
> {
>     for (auto& process : m_processes) {
>         if (process->canSendMessage() && !process->isPrewarmed() && process->sessionID() == sessionID)
>             process->send(T(message), 0);
>     }
> }
> ```
> 
> Note how it explicitly disregards prewarmed processes. I think we should explicitly ignore crashy processes too instead of returning an invalid ID (which is a new thing I'd rather not have to deal with).

I posted an updated PR with the explicit crash process check.

File: Source/WebKit/UIProcess/WebProcessProxy.h [hunk 1]
```diff
@@ -218,6 +218,10 @@ class WebProcessProxy : public AuxiliaryProcessProxy, private ProcessThrottlerCl
 
     bool isDummyProcessProxy() const;
 
+#if ENABLE(WEBCONTENT_CRASH_TESTING)
+    bool isCrashyProcess() { return m_isWebContentCrashyProcess; }
```
Comment (cdumez) [inline Source/WebKit/UIProcess/WebProcessProxy.h]: Function should be marked as const.

**Review by cdumez — APPROVED** [review]
LGTM with change.

File: Source/WebKit/UIProcess/WebProcessProxy.h (hunk 1 above)
Comment (msaboff) [reply Source/WebKit/UIProcess/WebProcessProxy.h]: Changed locally.

**Pull request merged**